llvm/include/llvm/Telemetry/Telemetry.h

```diff
@@ -0,0 +1,165 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides the basic framework for Telemetry.
+/// Refer to its documentation at llvm/docs/Telemetry.rst for more details.
+//===---------------------------------------------------------------------===//
+
+#ifndef LLVM_TELEMETRY_TELEMETRY_H
+#define LLVM_TELEMETRY_TELEMETRY_H
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+#include <map>
+#include <memory>
+#include <optional>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+namespace llvm {
+namespace telemetry {
+
+class Serializer {
+public:
+  virtual Error init() = 0;
+  virtual void write(StringRef KeyName, bool Value) = 0;
+  virtual void write(StringRef KeyName, int Value) = 0;
+  virtual void write(StringRef KeyName, unsigned int Value) = 0;
+  virtual void write(StringRef KeyName, unsigned long Value) = 0;
+  virtual void write(StringRef KeyName, long Value) = 0;
+  virtual void write(StringRef KeyName, long long Value) = 0;
+  virtual void write(StringRef KeyName, unsigned long long Value) = 0;
+  virtual void write(StringRef KeyName, StringRef Value) = 0;
+  virtual void beginObject(StringRef KeyName) = 0;
+  virtual void endObject() = 0;
```

Comment:
Question (for both of you) do you want begin/end object to be public, or should it be a private interface, accessible only through the map wrapper?

Reply:
I think it should be public to allow for nesting objects. 
Eg.,
```
s.beginObject("Parent");
s.beginObject("Child");
s.write("ChildAge", 100);
s.beginObject("GrandChild");
s.write("GrandKidAage", 10);
s.endObject();
s.endObject();
....
```

Reply:
Makes sense. Some of that could be achieved through types like `map<string, map<string, ??>>`, but this is definitely more flexible. (At the cost of potentially forgetting to call endObject, but if needed, we could solve that with a scoped RAII helper.)

Reply:
I have an assert in the `finalize()` call to make sure all children have been "ended" properly.